### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Matrix): Permanent

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib.lean [hunk 1]
```diff
@@ -3271,8 +3271,9 @@ import Mathlib.LinearAlgebra.Matrix.Nondegenerate
 import Mathlib.LinearAlgebra.Matrix.NonsingularInverse
 import Mathlib.LinearAlgebra.Matrix.Orthogonal
 import Mathlib.LinearAlgebra.Matrix.Permutation
-import Mathlib.LinearAlgebra.Matrix.Polynomial
+import Mathlib.LinearAlgebra.Matrix.Permanent
 import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.LinearAlgebra.Matrix.Permutation
```
Comment (mo271) [inline Mathlib.lean:3276]:
```suggestion
import Mathlib.LinearAlgebra.Matrix.Permanent
import Mathlib.LinearAlgebra.Matrix.Permutation
import Mathlib.LinearAlgebra.Matrix.PosDef
```

Comment (mo271) [reply Mathlib.lean]: fixing this manually

**Review by loefflerd — COMMENTED** [review]
This looks pretty good to me – I've suggested some localised changes and golfs but the basic design looks just fine (modulo the impression that the empty matrix is getting more attention than it deserves).

**Comment by loefflerd** [comment]
Sorry, those linter failures are my fault. VS Code strips trailing whitespace when you save, but I copy-pasted these lines without saving so there was apparently some rogue whitespace.

**Comment by mo271** [comment]
> Sorry, those linter failures are my fault. VS Code strips trailing whitespace when you save, but I copy-pasted these lines without saving so there was apparently some rogue whitespace.

no problem, I'll fix them...

**Comment by loefflerd** [comment]
I'm afraid you've got an unused import somewhere

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+* `Matrix.permanent`: the permanentanent of a square matrix, as a sum over permanentutations
+
+-/
+
+open Equiv Fintype Finset
+
+namespace Matrix
+
+open Matrix
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:24]: You don't need to open namespaces when you're already inside them. (Sorry, I missed this earlier.)

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+variable {n : Type*} [DecidableEq n] [Fintype n]
+variable {R : Type*} [CommSemiring R]
+
+/-- The permanentanent of a matrix defined as a sum over all permanentutations -/
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:29]:
```suggestion
/-- The permanent of a matrix defined as a sum over all permutations -/
```
:rofl: I like the idea of "permanentutations"

(there are a couple more of these in docstrings below)

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+/-- Transposing a matrix preserves the permanentanent. -/
+@[simp]
+theorem permanent_transpose (M : Matrix n n R) : Mᵀ.permanent = M.permanent := by
+  refine sum_bijective _ inv_involutive.bijective _ _ ?_
+  intro σ
+  apply Fintype.prod_equiv σ
+  simp
+
+/-- permuting the columns does not change the permanentanent. -/
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:77]:
```suggestion
/-- Permuting the columns does not change the permanent. -/
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+theorem permanent_permute_cols (σ : Perm n) (M : Matrix n n R) :
+    (M.submatrix σ id).permanent = M.permanent :=
+  (Group.mulLeft_bijective σ).sum_comp fun τ ↦ ∏ i : n, M (τ i) i
+
+/-- permuting the rows does not change the permanentanent. -/
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:82]:
```suggestion
/-- Permuting the rows does not change the permanent. -/
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    permanent A = A k k := by
+  have := uniqueOfSubsingleton k
+  convert permanent_unique A
+
+theorem permanent_eq_elem_of_card_eq_one {A : Matrix n n R} (h : card n = 1) (k : n) :
+    permanent A = A k k :=
+  haveI : Subsingleton n := card_le_one_iff_subsingleton.mp h.le
+  permanent_eq_elem_of_subsingleton _ _
+
+/-- Transposing a matrix preserves the permanentanent. -/
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:69]:
```suggestion
/-- Transposing a matrix preserves the permanent. -/
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+theorem permanent_one : permanent (1 : Matrix n n R) = 1 := by
+  rw [← diagonal_one]; simp [-diagonal_one]
+
+theorem permanent_isEmpty [IsEmpty n] {A : Matrix n n R} : permanent A = 1 := by simp [permanent]
+
+theorem permanent_eq_one_of_card_eq_zero {A : Matrix n n R} (h : card n = 0) : permanent A = 1 :=
+  haveI : IsEmpty n := card_eq_zero_iff.mp h
+  permanent_isEmpty
+
+/-- If `n` has only one element, the permanentanent of an `n` by `n` matrix is just that element.
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:52]:
```suggestion
/-- If `n` has only one element, the permanent of an `n` by `n` matrix is just that element.
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 8]
```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2024 Moritz Firsching. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching
+-/
+import Mathlib.Data.Fintype.Perm
+import Mathlib.Data.Matrix.Basic
+/-!
+# permanentanent of a matrix
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:10]:
```suggestion
# Permanent of a matrix
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+This file defines the permanentanent of a matrix, `Matrix.permanent`, and some of its properties.
+
+## Main definitions
+
+* `Matrix.permanent`: the permanentanent of a square matrix, as a sum over permanentutations
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:16]:
```suggestion
* `Matrix.permanent`: the permanent of a square matrix, as a sum over permutations
```

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+Copyright (c) 2024 Moritz Firsching. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching
+-/
+import Mathlib.Data.Fintype.Perm
+import Mathlib.Data.Matrix.Basic
+/-!
+# permanentanent of a matrix
+
+This file defines the permanentanent of a matrix, `Matrix.permanent`, and some of its properties.
```
Comment (loefflerd) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:12]:
```suggestion
This file defines the permanent of a matrix, `Matrix.permanent`, and some of its properties.
```

**Comment by mo271** [comment]
Thanks for fixing the regular-expresstion-sustitution gone wrong...

**Review by loefflerd — COMMENTED** [review]
All looks good now, thanks!

maintainer merge

File: Mathlib/LinearAlgebra/Matrix/Permanent.lean [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+-/
+
+open Equiv Fintype Finset
+
+namespace Matrix
+
+variable {n : Type*} [DecidableEq n] [Fintype n]
+variable {R : Type*} [CommSemiring R]
+
+/-- The permanent of a matrix defined as a sum over all permutations -/
```
Comment (urkud) [inline Mathlib/LinearAlgebra/Matrix/Permanent.lean:27]: Please make the docstring useful for people who don't know what's the permanent of a matrix. Same for the module docstring. Also, Wikipedia says that some authors define permanent of a rectangular matrix too. If you don't want to do it (I never used permanent in my research, so I have no opinion here), then please mention this choice in the module docstring.

Comment (urkud) [reply Mathlib/LinearAlgebra/Matrix/Permanent.lean:27]: Otherwise LGTM. Thanks!
bors d+

Comment (urkud) [reply Mathlib/LinearAlgebra/Matrix/Permanent.lean:27]: Optional: add
- `permanent_smul_col`, `permanent_smul_row`, `permanent_smul`
- add `permanent` as a `MultilinearMap`. Do we have symmetric multilinear maps?

Comment (mo271) [reply Mathlib/LinearAlgebra/Matrix/Permanent.lean]: I added a sentence explaining the relation to determinant; this way the docstring is more useful for people who already know what a determinant is. Also adding `square` in order to make sure that this is not the more exotic more general rectangular permanent. 

Regarding the theorems/definitions you mention: I agree they should be added, but let's do it in another PR.

Comment (mo271) [reply Mathlib/LinearAlgebra/Matrix/Permanent.lean]: > Optional (can wait for a future PR): add
> 
> * `permanent_smul_col`, `permanent_smul_row`, `permanent_smul`

Adding those in #19307

**Pull request closed without merging**